include/range/v3/utility/variant.hpp

```diff
@@ -251,12 +251,13 @@ namespace ranges
                 apply_visitor(Fun &&fun, Var &var)
                   : var_(var), fun_(std::forward<Fun>(fun))
                 {}
-                template<typename T, std::size_t N>
-                void operator()(T &&t, meta::size_t<N> u) const
-                {
-                    using result_t = result_of_t<Fun(T &&, meta::size_t<N>)>;
-                    this->apply_(std::forward<T>(t), u, std::is_void<result_t>{});
-                }
+                template<typename T, std::size_t N,
+                         typename result_t = result_of_t<Fun(T &&, meta::size_t<N>)>>
+                auto operator()(T &&t, meta::size_t<N> u) const
+                RANGES_DECLTYPE_AUTO_RETURN
+                (
+                    this->apply_(std::forward<T>(t), u, std::is_void<result_t>{})
```

Comment:
Hm I do not recall anything in particular about gcc and this but in the
constexpr algorithms branch the only tests that fail are gcc because it
somehow doesn't like the constrained implementation of move :/
I really don't know what is wrong there.

On Wednesday, May 13, 2015, Eric Niebler <notifications@github.com> wrote:

> In include/range/v3/utility/variant.hpp
> <https://github.com/ericniebler/range-v3/pull/145#discussion_r30194152>:
>
> > @@ -251,12 +251,13 @@ namespace ranges
> >                  apply_visitor(Fun &&fun, Var &var)
> >                    : var_(var), fun_(std::forward<Fun>(fun))
> >                  {}
> > -                template<typename T, std::size_t N>
> > -                void operator()(T &&t, meta::size_t<N> u) const
> > -                {
> > -                    using result_t = result_of_t<Fun(T &&, meta::size_t<N>)>;
> > -                    this->apply_(std::forward<T>(t), u, std::is_void<result_t>{});
> > -                }
> > +                template<typename T, std::size_t N,
> > +                         typename result_t = result_of_t<Fun(T &&, meta::size_t<N>)>>
> > +                auto operator()(T &&t, meta::size_t<N> u) const
> > +                RANGES_DECLTYPE_AUTO_RETURN
> > +                (
> > +                    this->apply_(std::forward<T>(t), u, std::is_void<result_t>{})
>
> I seem to recall that gcc has a hard time with this appearing in contexts
> like these. Apparently not since the tests passed. ???
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ericniebler/range-v3/pull/145/files#r30194152>.
>